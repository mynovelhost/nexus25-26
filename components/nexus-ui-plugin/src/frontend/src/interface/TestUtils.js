/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import {fireEvent, render, wait} from '@testing-library/react';
import UIStrings from '../constants/UIStrings';

/**
 * @since 3.24
 */
export default class Utils {
  static render(view, extraSelectors) {
    const selectors = render(view);
    const {queryByText} = selectors;
    return {
      ...selectors,
      loadingMask: () => queryByText('Loadingâ€¦'),
      savingMask: () => queryByText(UIStrings.SAVING),
      ...extraSelectors(selectors)};
  }

  static async changeField(fieldSelector, value) {
    fireEvent.change(fieldSelector(), {
      target: {
        name: fieldSelector().name,
        value
      }
    });
    await wait(() => expect(fieldSelector()).toHaveValue(value));
  }
}
