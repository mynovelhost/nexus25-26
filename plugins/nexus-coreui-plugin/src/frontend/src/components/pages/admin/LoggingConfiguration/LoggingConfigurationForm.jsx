/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import {useMachine} from '@xstate/react';

import {
  Button, ContentBody,
  FieldWrapper,
  NxErrorAlert,
  NxLoadWrapper,
  NxSubmitMask,
  Page,
  PageHeader,
  PageTitle,
  Section,
  SectionFooter,
  Select,
  Textfield,
  Utils
} from 'nexus-ui-plugin';

import LoggingConfigurationFormMachine from './LoggingConfigurationFormMachine';

import UIStrings from '../../../../constants/UIStrings';
import {faScroll} from '@fortawesome/free-solid-svg-icons';

export default function LoggingConfigurationForm({itemId, onDone}) {
  const [current, send] = useMachine(LoggingConfigurationFormMachine, {
    context: {
      pristineData: {
        name: itemId
      }
    },

    actions: {
      onCancel: onDone,
      onReset: onDone,
      onSaveSuccess: onDone
    },

    devTools: true
  });

  const {isPristine, pristineData, data, loadError, saveError, validationErrors} = current.context;
  const isLoading = current.matches('loading');
  const isSaving = current.matches('saving');
  const isResetting = current.matches('resetting');
  const isInvalid = Utils.isInvalid(validationErrors);
  const hasData = data && data !== {};

  function update(event) {
    send('UPDATE', {data: {[event.target.name]: event.target.value}});
  }

  function save(event) {
    event.preventDefault();
    send('SAVE');
  }

  function handleEnter(event) {
    if (event.key === 'Enter') {
      save(event);
    }
  }

  function cancel() {
    send('CANCEL');
  }

  function reset() {
    send('RESET');
  }

  return <Page>
    <PageHeader><PageTitle icon={faScroll} {...UIStrings.LOGGING.MENU}/></PageHeader>
    <ContentBody className="nxrm-logging-configuration">
      <Section className="nxrm-logging-configuration-form" onKeyPress={handleEnter}>
        <NxLoadWrapper loading={isLoading} error={loadError ? `${loadError}` : null}>
      {hasData && <>
        {saveError && <NxErrorAlert>{UIStrings.LOGGING.MESSAGES.SAVE_ERROR} {saveError}</NxErrorAlert>}
        {isSaving && <NxSubmitMask message={UIStrings.SAVING}/>}
        {isResetting && <NxSubmitMask message={UIStrings.LOGGING.MESSAGES.RESETTING}/>}

        <FieldWrapper labelText={UIStrings.LOGGING.NAME_LABEL}>
          <Textfield
              name="name"
              disabled={pristineData.name}
              value={data.name}
              onChange={update}
              validationErrors={validationErrors.name}/>
        </FieldWrapper>
        <FieldWrapper labelText={UIStrings.LOGGING.LEVEL_LABEL}>
          <Select name="level" value={data.level} onChange={update}>
            {['OFF', 'ERROR', 'WARN', 'INFO', 'DEBUG', 'TRACE'].map(logLevel =>
                <option key={logLevel} value={logLevel}>{logLevel}</option>
            )}
          </Select>
        </FieldWrapper>

        <SectionFooter>
          <Button variant="primary" disabled={isPristine || isInvalid} onClick={save} type="submit">
            {UIStrings.SETTINGS.SAVE_BUTTON_LABEL}
          </Button>
          <Button onClick={cancel}>{UIStrings.SETTINGS.CANCEL_BUTTON_LABEL}</Button>
          {itemId && <Button variant="error" onClick={reset}>{UIStrings.LOGGING.RESET_BUTTON}</Button>}
        </SectionFooter>
      </>}
        </NxLoadWrapper>
      </Section>
    </ContentBody>
  </Page>;
}
