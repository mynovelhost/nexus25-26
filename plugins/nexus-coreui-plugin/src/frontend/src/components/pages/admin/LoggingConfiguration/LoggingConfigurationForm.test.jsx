/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import {fireEvent, wait, waitForElement, waitForElementToBeRemoved} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TestUtils from 'nexus-ui-plugin/src/frontend/src/interface/TestUtils';
import axios from 'axios';
import {ExtJS} from 'nexus-ui-plugin';

import LoggingConfigurationForm from './LoggingConfigurationForm';

import UIStrings from "../../../../constants/UIStrings";

jest.mock('axios', () => ({
  ...jest.requireActual('axios'), // Use most functions from actual axios
  get: jest.fn(),
  put: jest.fn(),
  post: jest.fn()
}));

jest.mock('nexus-ui-plugin', () => ({
  ...jest.requireActual('nexus-ui-plugin'),
  ExtJS: {
    requestConfirmation: jest.fn()
  },
  Utils: {
    buildFormMachine: function(args) {
      const machine = jest.requireActual('nexus-ui-plugin').Utils.buildFormMachine(args);
      return machine.withConfig({
        actions: {
          logSaveSuccess: jest.fn(),
          logLoadError: jest.fn()
        }
      })
    },
    isInvalid: jest.requireActual('nexus-ui-plugin').Utils.isInvalid,
    isBlank: jest.requireActual('nexus-ui-plugin').Utils.isBlank,
    notBlank: jest.requireActual('nexus-ui-plugin').Utils.notBlank
  }
}));

describe('LoggingConfigurationForm', function() {
  const CONFIRM = Promise.resolve();
  const OVERRIDDEN_LOGGER = Promise.resolve({data: {override: true}});
  const onDone = jest.fn();

  function renderEditView() {
    return renderView(<LoggingConfigurationForm itemId="ROOT" onDone={onDone}/>);
  }

  function renderCreateView() {
    return renderView(<LoggingConfigurationForm onDone={onDone} />);
  }

  function renderView(view) {
    return TestUtils.render(view, ({queryByLabelText, queryByText}) => ({
      name: () => queryByLabelText(UIStrings.LOGGING.NAME_LABEL),
      level: () => queryByLabelText(UIStrings.LOGGING.LEVEL_LABEL),
      saveButton: () => queryByText(UIStrings.SETTINGS.SAVE_BUTTON_LABEL),
      cancelButton: () => queryByText(UIStrings.SETTINGS.CANCEL_BUTTON_LABEL),
      resetMask: () => queryByText(UIStrings.LOGGING.MESSAGES.RESETTING),
      resetButton: () => queryByText(UIStrings.LOGGING.RESET_BUTTON)
    }));
  }

  const changeFieldAndAssertValue = async (fieldSelector, value) => {
    fireEvent.change(fieldSelector(), {target: {value: value}});
    await wait(() => expect(fieldSelector()).toHaveValue(value));
  };

  it('renders a loading spinner', async function() {
    axios.get.mockReturnValue(new Promise(() => {}));

    const {container, loadingMask} = renderEditView();

    await waitForElement(loadingMask);

    expect(container).toMatchSnapshot();
  });

  it('renders the resolved data', async function() {
    axios.get.mockResolvedValue({
      data: {name: 'ROOT', level: 'INFO'}
    });

    const {container, loadingMask, name, level, saveButton} = renderEditView();

    await waitForElementToBeRemoved(loadingMask);

    expect(name()).toHaveValue('ROOT');
    expect(level()).toHaveValue('INFO');
    expect(saveButton()).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it('renders an error message', async function() {
    axios.get.mockImplementation(() => new Promise((resolve, reject) => {
      setTimeout(() => reject({message: 'Error'}), 100);
    }));

    const {container, loadingMask} = renderEditView();

    await waitForElementToBeRemoved(loadingMask);

    expect(container.querySelector('.nx-alert--error')).toHaveTextContent('Error');
  });

  it('requires the name field when creating a new logging configuration', async function() {
    const {container, loadingMask, name, level, saveButton} = renderCreateView();

    await waitForElementToBeRemoved(loadingMask);

    expect(container).toMatchSnapshot();

    expect(saveButton()).toBeDisabled();
    expect(level()).toHaveValue('INFO');

    await changeFieldAndAssertValue(name, 'name');

    expect(saveButton()).toBeEnabled();
  });

  it('fires onDone when cancelled', async function() {
    const {loadingMask, cancelButton} = renderCreateView();

    await waitForElementToBeRemoved(loadingMask);

    fireEvent.click(cancelButton());

    await wait(() => expect(onDone).toBeCalled());
  });

  it('requests confirmation when the logger is overridden and saves when requested', async function() {
    const {loadingMask, name, level, saveButton} = renderCreateView();

    await waitForElementToBeRemoved(loadingMask);

    axios.put.mockReturnValue(Promise.resolve());

    await changeFieldAndAssertValue(name, 'name');
    await changeFieldAndAssertValue(level, 'DEBUG');
    expect(saveButton()).toBeEnabled();

    axios.get.mockReturnValue(OVERRIDDEN_LOGGER);
    ExtJS.requestConfirmation.mockReturnValue(CONFIRM);
    fireEvent.click(saveButton());

    await wait(() => expect(axios.put).toBeCalledWith(
        '/service/rest/internal/ui/loggingConfiguration/name',
        {name: 'name', level: 'DEBUG'}
    ));
    expect(onDone).toBeCalled();
  });

  it('saves', async function() {
    axios.get.mockReturnValue(Promise.resolve({
      data: {name: 'ROOT', level: 'INFO'}
    }));

    const {loadingMask, level, saveButton} = renderEditView();

    await waitForElementToBeRemoved(loadingMask);

    await wait(() => expect(window.dirty).toEqual([]));

    await changeFieldAndAssertValue(level, 'DEBUG');

    await wait(() => expect(window.dirty).toEqual(['LoggingConfigurationFormMachine']));

    fireEvent.click(saveButton());

    await wait(() => expect(axios.put).toHaveBeenLastCalledWith(
        '/service/rest/internal/ui/loggingConfiguration/ROOT',
        {name: 'ROOT', level: 'DEBUG'}
    ));
    expect(window.dirty).toEqual([]);
  });

  it('resets an edited logger', async function() {
    axios.get.mockReturnValue(Promise.resolve({
      data: {name: 'test', level: 'ERROR'}
    }));

    const {loadingMask, resetButton, resetMask} = renderEditView();

    await waitForElementToBeRemoved(loadingMask);

    ExtJS.requestConfirmation.mockReturnValue(CONFIRM);
    axios.post.mockReturnValue(CONFIRM);
    fireEvent.click(resetButton());

    await wait(() => expect(resetMask()).toBeInTheDocument());

    expect(axios.post).toBeCalledWith('/service/rest/internal/ui/loggingConfiguration/test/reset');
    expect(onDone).toBeCalled();
  });
});
