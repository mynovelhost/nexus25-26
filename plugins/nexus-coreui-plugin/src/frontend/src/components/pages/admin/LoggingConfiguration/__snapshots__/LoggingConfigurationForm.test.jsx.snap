// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoggingConfigurationForm renders a loading spinner 1`] = `
<div>
  <div
    class="nxrm-page"
  >
    <div
      class="nxrm-page-header"
    >
      <div
        class="nxrm-page-title"
      >
        <h1
          class="title"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-scroll fa-w-20 nx-icon"
            data-icon="scroll"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M48 0C21.53 0 0 21.53 0 48v64c0 8.84 7.16 16 16 16h80V48C96 21.53 74.47 0 48 0zm208 412.57V352h288V96c0-52.94-43.06-96-96-96H111.59C121.74 13.41 128 29.92 128 48v368c0 38.87 34.65 69.65 74.75 63.12C234.22 474 256 444.46 256 412.57zM288 384v32c0 52.93-43.06 96-96 96h336c61.86 0 112-50.14 112-112 0-8.84-7.16-16-16-16H288z"
              fill="currentColor"
            />
          </svg>
          <span
            class="nx-feature-name"
          >
            Logging
          </span>
        </h1>
        <p
          class="description nx-feature-description"
        >
          Control logging levels
        </p>
      </div>
    </div>
    <div
      class="nxrm-content-body nxrm-logging-configuration"
    >
      <div
        class="nxrm-section nx-tile nxrm-logging-configuration-form"
      >
        <div
          class="nx-loading-spinner"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-circle-notch fa-w-16 fa-spin nx-loading-spinner__icon nx-icon"
            data-icon="circle-notch"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z"
              fill="currentColor"
            />
          </svg>
          <span>
            Loadingâ€¦
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoggingConfigurationForm renders the resolved data 1`] = `
<div>
  <div
    class="nxrm-page"
  >
    <div
      class="nxrm-page-header"
    >
      <div
        class="nxrm-page-title"
      >
        <h1
          class="title"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-scroll fa-w-20 nx-icon"
            data-icon="scroll"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M48 0C21.53 0 0 21.53 0 48v64c0 8.84 7.16 16 16 16h80V48C96 21.53 74.47 0 48 0zm208 412.57V352h288V96c0-52.94-43.06-96-96-96H111.59C121.74 13.41 128 29.92 128 48v368c0 38.87 34.65 69.65 74.75 63.12C234.22 474 256 444.46 256 412.57zM288 384v32c0 52.93-43.06 96-96 96h336c61.86 0 112-50.14 112-112 0-8.84-7.16-16-16-16H288z"
              fill="currentColor"
            />
          </svg>
          <span
            class="nx-feature-name"
          >
            Logging
          </span>
        </h1>
        <p
          class="description nx-feature-description"
        >
          Control logging levels
        </p>
      </div>
    </div>
    <div
      class="nxrm-content-body nxrm-logging-configuration"
    >
      <div
        class="nxrm-section nx-tile nxrm-logging-configuration-form"
      >
        <div
          class="field-wrapper"
        >
          <label
            class="field-wrapper-label"
            for="name"
          >
            Logger Name
          </label>
          <input
            class="nxrm-textfield"
            disabled=""
            id="name"
            name="name"
            type="text"
            value="ROOT"
          />
        </div>
        <div
          class="field-wrapper"
        >
          <label
            class="field-wrapper-label"
            for="level"
          >
            Logger Level
          </label>
          <select
            class="nxrm-select"
            id="level"
            name="level"
          >
            <option
              value="OFF"
            >
              OFF
            </option>
            <option
              value="ERROR"
            >
              ERROR
            </option>
            <option
              value="WARN"
            >
              WARN
            </option>
            <option
              value="INFO"
            >
              INFO
            </option>
            <option
              value="DEBUG"
            >
              DEBUG
            </option>
            <option
              value="TRACE"
            >
              TRACE
            </option>
          </select>
        </div>
        <div
          class="nxrm-section-footer nx-tile-footer"
        >
          <button
            class="nx-btn nx-btn--primary"
            disabled=""
            type="submit"
          >
            Save
          </button>
          <button
            class="nx-btn"
            type="input"
          >
            Cancel
          </button>
          <button
            class="nx-btn nx-btn--error"
            type="input"
          >
            Reset to Default Level
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoggingConfigurationForm requires the name field when creating a new logging configuration 1`] = `
<div>
  <div
    class="nxrm-page"
  >
    <div
      class="nxrm-page-header"
    >
      <div
        class="nxrm-page-title"
      >
        <h1
          class="title"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-scroll fa-w-20 nx-icon"
            data-icon="scroll"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M48 0C21.53 0 0 21.53 0 48v64c0 8.84 7.16 16 16 16h80V48C96 21.53 74.47 0 48 0zm208 412.57V352h288V96c0-52.94-43.06-96-96-96H111.59C121.74 13.41 128 29.92 128 48v368c0 38.87 34.65 69.65 74.75 63.12C234.22 474 256 444.46 256 412.57zM288 384v32c0 52.93-43.06 96-96 96h336c61.86 0 112-50.14 112-112 0-8.84-7.16-16-16-16H288z"
              fill="currentColor"
            />
          </svg>
          <span
            class="nx-feature-name"
          >
            Logging
          </span>
        </h1>
        <p
          class="description nx-feature-description"
        >
          Control logging levels
        </p>
      </div>
    </div>
    <div
      class="nxrm-content-body nxrm-logging-configuration"
    >
      <div
        class="nxrm-section nx-tile nxrm-logging-configuration-form"
      >
        <div
          class="field-wrapper"
        >
          <label
            class="field-wrapper-label"
            for="name"
          >
            Logger Name
          </label>
          <input
            class="nxrm-textfield invalid"
            id="name"
            name="name"
            type="text"
            value=""
          />
          <span
            class="nxrm-error-message"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-exclamation-circle fa-w-16 nx-icon"
              data-icon="exclamation-circle"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                fill="currentColor"
              />
            </svg>
            <span
              class="nxrm-error-message-text"
              tabindex="-1"
            >
               
              This field is required
               
            </span>
          </span>
        </div>
        <div
          class="field-wrapper"
        >
          <label
            class="field-wrapper-label"
            for="level"
          >
            Logger Level
          </label>
          <select
            class="nxrm-select"
            id="level"
            name="level"
          >
            <option
              value="OFF"
            >
              OFF
            </option>
            <option
              value="ERROR"
            >
              ERROR
            </option>
            <option
              value="WARN"
            >
              WARN
            </option>
            <option
              value="INFO"
            >
              INFO
            </option>
            <option
              value="DEBUG"
            >
              DEBUG
            </option>
            <option
              value="TRACE"
            >
              TRACE
            </option>
          </select>
        </div>
        <div
          class="nxrm-section-footer nx-tile-footer"
        >
          <button
            class="nx-btn nx-btn--primary"
            disabled=""
            type="submit"
          >
            Save
          </button>
          <button
            class="nx-btn"
            type="input"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
